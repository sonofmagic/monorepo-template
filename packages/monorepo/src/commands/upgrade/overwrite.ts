import { Buffer } from 'node:buffer'
import checkbox from '@inquirer/checkbox'
import fs from 'fs-extra'
import pc from 'picocolors'
import { isFileChanged } from '../../utils'

type WriteIntent
  = (
      { type: 'write', reason: 'missing' }
      | { type: 'skip', reason: 'skipOverwrite' | 'identical' }
      | { type: 'prompt', reason: 'changed' }
  )

function asBuffer(data: Buffer | string) {
  return typeof data === 'string' ? Buffer.from(data) : data
}

export async function evaluateWriteIntent(
  targetPath: string,
  options: {
    skipOverwrite?: boolean
    source: Buffer | string
  },
): Promise<WriteIntent> {
  const { skipOverwrite, source } = options
  const exists = await fs.pathExists(targetPath)
  if (!exists) {
    return {
      type: 'write',
      reason: 'missing',
    }
  }

  if (skipOverwrite) {
    return {
      type: 'skip',
      reason: 'skipOverwrite',
    }
  }

  const src = asBuffer(source)
  let destSize = 0
  try {
    const stat = await fs.stat(targetPath)
    destSize = stat.size
  }
  catch {
    return {
      type: 'write',
      reason: 'missing',
    }
  }

  if (destSize !== src.length) {
    return {
      type: 'prompt',
      reason: 'changed',
    }
  }

  const dest = await fs.readFile(targetPath)
  if (!isFileChanged(src, dest)) {
    return {
      type: 'skip',
      reason: 'identical',
    }
  }

  return {
    type: 'prompt',
    reason: 'changed',
  }
}

export interface PendingOverwrite {
  relPath: string
  targetPath: string
  action: () => Promise<void>
}

export async function scheduleOverwrite(
  intent: WriteIntent,
  options: {
    relPath: string
    targetPath: string
    action: () => Promise<void>
    pending: PendingOverwrite[]
  },
) {
  const { relPath, targetPath, action, pending } = options

  if (intent.type === 'write') {
    await action()
    return
  }

  if (intent.type === 'prompt') {
    pending.push({
      relPath,
      targetPath,
      action,
    })
  }
}

export async function flushPendingOverwrites(pending: PendingOverwrite[]) {
  if (!pending.length) {
    return
  }

  const selected = await checkbox({
    message: '检测到以下文件内容与当前仓库不同，选择需要覆盖的文件',
    choices: pending.map(item => ({
      name: pc.greenBright(item.relPath),
      value: item.targetPath,
      checked: false,
    })),
    loop: false,
  })

  const selectedSet = new Set(selected)
  for (const item of pending) {
    if (selectedSet.has(item.targetPath)) {
      await item.action()
    }
  }
}
