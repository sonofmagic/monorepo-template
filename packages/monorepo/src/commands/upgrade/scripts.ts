/**
 * 升级时注入到 package.json 的脚本命令集合，保证常用脚本齐全。
 */
export const scripts = {
  'script:init': 'monorepo init',
  'script:sync': 'monorepo sync',
  'script:clean': 'monorepo clean',
  'script:mirror': 'monorepo mirror',
  'commitlint': 'commitlint --edit',
}

export const scriptsEntries = Object.entries(scripts)
