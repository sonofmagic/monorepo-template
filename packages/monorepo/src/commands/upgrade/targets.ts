export function getAssetTargets(raw?: boolean) {
  const list: string[] = [
    '.changeset',
    '.husky',
    '.vscode',
    '.editorconfig',
    '.gitattributes',
    '.gitignore',
    '.npmrc',
    'commitlint.config.ts',
    'eslint.config.js',
    'lint-staged.config.js',
    'stylelint.config.js',
    'package.json',
    // pnpm
    'pnpm-workspace.yaml',
    // base tsconfig
    'tsconfig.json',
    // turbo
    'turbo.json',
    // vitest
    'vitest.config.ts',
    // 'vitest.workspace.ts',
    // #region docker
    'Dockerfile',
    '.dockerignore',
    // #endregion
  ]
  if (!raw) {
    // renovate
    list.push('.github', 'LICENSE', 'renovate.json', 'SECURITY.md', 'CODE_OF_CONDUCT.md', 'CONTRIBUTING.md', 'netlify.toml')
  }
  return list
}
