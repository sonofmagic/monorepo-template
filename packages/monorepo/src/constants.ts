import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { name, version } from '../package.json'

export {
  name,
  version,
}

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
// import.meta.dirname for Nodejs >= v20.11.0
// https://nodejs.org/api/esm.html#importmetadirname

export const templatesDir = path.join(__dirname, '../templates')

export const assetsDir = path.join(__dirname, '../assets')

export const rootDir = path.join(__dirname, '../../..')
