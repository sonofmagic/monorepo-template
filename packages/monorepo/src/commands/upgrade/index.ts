import type { CliOpts, PackageJson } from '../../types'
import { Buffer } from 'node:buffer'
import process from 'node:process'
import checkbox from '@inquirer/checkbox'
import confirm from '@inquirer/confirm'
import fs from 'fs-extra'
import klaw from 'klaw'
import path from 'pathe'
import pc from 'picocolors'
import { coerce, gte, minVersion } from 'semver'
import set from 'set-value'
import { assetsDir, name as pkgName, version as pkgVersion } from '../../constants'
import { resolveCommandConfig } from '../../core/config'
import { GitClient } from '../../core/git'
import { logger } from '../../core/logger'
import { escapeStringRegexp, isFileChanged, isIgnorableFsError, isMatch } from '../../utils'
import { scriptsEntries } from './scripts'
import { getAssetTargets } from './targets'

function isWorkspace(version?: string) {
  if (typeof version === 'string') {
    return version.startsWith('workspace:')
  }
  return false
}

function parseVersion(input: unknown) {
  if (typeof input !== 'string' || input.trim().length === 0) {
    return null
  }
  return minVersion(input) ?? coerce(input)
}

function shouldAssignVersion(currentVersion: unknown, nextVersion: string) {
  if (typeof currentVersion !== 'string' || currentVersion.trim().length === 0) {
    return true
  }

  if (currentVersion === nextVersion) {
    return false
  }

  const current = parseVersion(currentVersion)
  const next = parseVersion(nextVersion)
  if (!current || !next) {
    return true
  }

  return !gte(current, next)
}

/**
 * 将内置 package.json 内容合并进目标工程：
 * - 同步依赖（保留 workspace: 前缀的版本）
 * - 确保 @icebreakers/monorepo 使用最新版本
 * - 写入预置脚本
 */
export function setPkgJson(
  sourcePkgJson: PackageJson,
  targetPkgJson: PackageJson,
  options?: {
    scripts?: Record<string, string>
  },
) {
  const packageManager = sourcePkgJson.packageManager ?? ''
  const sourceDeps = sourcePkgJson.dependencies ?? {}
  const sourceDevDeps = sourcePkgJson.devDependencies ?? {}

  const targetDeps = { ...(targetPkgJson.dependencies ?? {}) }
  const targetDevDeps = { ...(targetPkgJson.devDependencies ?? {}) }

  if (packageManager) {
    targetPkgJson.packageManager = packageManager
  }

  for (const [depName, depVersion] of Object.entries(sourceDeps)) {
    if (typeof depVersion !== 'string') {
      continue
    }

    const targetVersion = targetDeps[depName]
    if (isWorkspace(targetVersion)) {
      continue
    }
    if (shouldAssignVersion(targetVersion, depVersion)) {
      targetDeps[depName] = depVersion
    }
  }
  if (Object.keys(targetDeps).length) {
    targetPkgJson.dependencies = targetDeps
  }

  for (const [depName, depVersion] of Object.entries(sourceDevDeps)) {
    if (typeof depVersion !== 'string') {
      continue
    }

    if (depName === pkgName) {
      const nextVersion = `^${pkgVersion}`
      const targetVersion = targetDevDeps[depName]
      if (!isWorkspace(targetVersion) && shouldAssignVersion(targetVersion, nextVersion)) {
        targetDevDeps[depName] = nextVersion
      }
    }
    else {
      const targetVersion = targetDevDeps[depName]
      if (isWorkspace(targetVersion)) {
        continue
      }
      if (shouldAssignVersion(targetVersion, depVersion)) {
        targetDevDeps[depName] = depVersion
      }
    }
  }
  if (Object.keys(targetDevDeps).length) {
    targetPkgJson.devDependencies = targetDevDeps
  }

  const scriptPairs = options?.scripts ? Object.entries(options.scripts) : scriptsEntries
  if (scriptPairs.length) {
    const scripts = { ...(targetPkgJson.scripts ?? {}) }
    for (const [scriptName, scriptCmd] of scriptPairs) {
      scripts[scriptName] = scriptCmd
    }
    targetPkgJson.scripts = scripts
  }
}

function confirmOverwrite(filename: string) {
  return confirm({ message: `${pc.greenBright(filename)} 文件内容发生改变,是否覆盖?`, default: true })
}

function asBuffer(data: Buffer | string) {
  return typeof data === 'string' ? Buffer.from(data) : data
}

/**
 * 判断是否需要写入目标文件，包含以下策略：
 * 1. 目标文件不存在：直接写入
 * 2. 开启 skipOverwrite：完全跳过
 * 3. 文件大小不同或内容 hash 不一致，提示用户确认
 */
async function shouldWriteFile(
  targetPath: string,
  options: {
    skipOverwrite?: boolean
    source: Buffer | string
    promptLabel: string
  },
) {
  const { skipOverwrite, source, promptLabel } = options
  const exists = await fs.pathExists(targetPath)
  if (!exists) {
    return true
  }

  if (skipOverwrite) {
    return false
  }

  const src = asBuffer(source)
  let destSize = 0
  try {
    const stat = await fs.stat(targetPath)
    destSize = stat.size
  }
  catch {
    return true
  }

  if (destSize !== src.length) {
    return confirmOverwrite(promptLabel)
  }

  const dest = await fs.readFile(targetPath)
  if (!isFileChanged(src, dest)) {
    return false
  }

  return confirmOverwrite(promptLabel)
}

/**
 * 将 assets 目录的模版文件同步到工程中，实现一键升级脚手架能力。
 */
export async function upgradeMonorepo(opts: CliOpts) {
  const cwd = opts.cwd ?? process.cwd()
  const upgradeConfig = await resolveCommandConfig('upgrade', cwd)
  const merged: CliOpts = {
    cwd,
    outDir: '',
    ...(upgradeConfig ?? {}),
    ...opts,
  }

  const outDir = merged.outDir ?? ''
  const absOutDir = path.isAbsolute(outDir) ? outDir : path.join(cwd, outDir)
  const gitClient = new GitClient({
    baseDir: cwd,
  })
  const repoName = await gitClient.getRepoName()
  // 默认从 assets 目录读取一组标准文件作为升级目标。
  const baseTargets = getAssetTargets(merged.raw)
  const configTargets = upgradeConfig?.targets ?? []
  const mergeTargets = upgradeConfig?.mergeTargets
  let targets = configTargets.length
    ? (mergeTargets === false ? [...configTargets] : Array.from(new Set([...baseTargets, ...configTargets])))
    : baseTargets

  if (merged.interactive) {
    // 交互模式允许用户临时调整需要覆盖的文件集合。
    // https://github.com/pnpm/pnpm/blob/db420ab592666dbae77fdda3f5c04ed2c045846d/pkg-manager/plugin-commands-installation/src/update/index.ts
    targets = await checkbox({
      message: '选择你需要的文件',
      choices: targets.map((x) => {
        return {
          value: x,
          checked: true,
        }
      }),
    })
  }

  const regexpArr = targets.map((x) => {
    return new RegExp(`^${escapeStringRegexp(x)}`)
  })
  // 旧版本默认跳过 changeset Markdown，可通过配置覆盖。
  const skipChangesetMarkdown = upgradeConfig?.skipChangesetMarkdown ?? true
  const scriptOverrides = upgradeConfig?.scripts
  const skipOverwrite = merged.skipOverwrite
  for await (const file of klaw(assetsDir, {
    filter(p) {
      const str = path.relative(assetsDir, p)
      return isMatch(str, regexpArr)
    },
  })) {
    if (!file.stats.isFile()) {
      continue
    }

    let relPath = path.relative(assetsDir, file.path)
    if (relPath === 'gitignore') {
      relPath = '.gitignore'
    }

    if (skipChangesetMarkdown && relPath.startsWith('.changeset/') && relPath.endsWith('.md')) {
      continue
    }
    const targetPath = path.resolve(absOutDir, relPath)

    try {
      if (relPath === 'package.json') {
        if (!await fs.pathExists(targetPath)) {
          continue
        }

        const sourcePkgJson = await fs.readJson(file.path) as PackageJson
        const targetPkgJson = await fs.readJson(targetPath) as PackageJson
        setPkgJson(sourcePkgJson, targetPkgJson, { scripts: scriptOverrides })
        // 直接覆写对象后重新序列化，保证键顺序与缩进一致。
        const data = `${JSON.stringify(targetPkgJson, undefined, 2)}\n`
        if (await shouldWriteFile(targetPath, { skipOverwrite, source: data, promptLabel: relPath })) {
          await fs.outputFile(targetPath, data, 'utf8')
          logger.success(targetPath)
        }
        continue
      }

      if (relPath === '.changeset/config.json' && repoName) {
        const changesetJson = await fs.readJson(file.path)
        set(changesetJson, 'changelog.1.repo', repoName)
        const data = `${JSON.stringify(changesetJson, undefined, 2)}\n`
        if (await shouldWriteFile(targetPath, { skipOverwrite, source: data, promptLabel: relPath })) {
          await fs.outputFile(targetPath, data, 'utf8')
          logger.success(targetPath)
        }
        continue
      }

      if (relPath === 'LICENSE') {
        const source = await fs.readFile(file.path)
        if (await shouldWriteFile(targetPath, { skipOverwrite: true, source, promptLabel: relPath })) {
          await fs.copy(file.path, targetPath)
          logger.success(targetPath)
        }
        continue
      }

      if (await shouldWriteFile(targetPath, { skipOverwrite, source: await fs.readFile(file.path), promptLabel: relPath })) {
        await fs.copy(file.path, targetPath)
        logger.success(targetPath)
      }
    }
    catch (error) {
      if (isIgnorableFsError(error)) {
        continue
      }
      throw error
    }
  }
}
