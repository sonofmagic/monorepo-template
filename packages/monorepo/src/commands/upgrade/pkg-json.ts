import type { PackageJson } from '../../types'
import { coerce, gte, minVersion } from 'semver'
import { name as pkgName, version as pkgVersion } from '../../constants'
import { scriptsEntries } from './scripts'

function isWorkspace(version?: string) {
  if (typeof version === 'string') {
    return version.startsWith('workspace:')
  }
  return false
}

function parseVersion(input: unknown) {
  if (typeof input !== 'string' || input.trim().length === 0) {
    return null
  }
  return minVersion(input) ?? coerce(input)
}

function shouldAssignVersion(currentVersion: unknown, nextVersion: string) {
  if (typeof currentVersion !== 'string' || currentVersion.trim().length === 0) {
    return true
  }

  if (currentVersion === nextVersion) {
    return false
  }

  const current = parseVersion(currentVersion)
  const next = parseVersion(nextVersion)
  if (!current || !next) {
    return true
  }

  return !gte(current, next)
}

/**
 * 将内置 package.json 内容合并进目标工程：
 * - 同步依赖（保留 workspace: 前缀的版本）
 * - 确保 @icebreakers/monorepo 使用最新版本
 * - 写入预置脚本
 */
export function setPkgJson(
  sourcePkgJson: PackageJson,
  targetPkgJson: PackageJson,
  options?: {
    scripts?: Record<string, string>
  },
) {
  const packageManager = sourcePkgJson.packageManager ?? ''
  const sourceDeps = sourcePkgJson.dependencies ?? {}
  const sourceDevDeps = sourcePkgJson.devDependencies ?? {}

  const targetDeps = { ...(targetPkgJson.dependencies ?? {}) }
  const targetDevDeps = { ...(targetPkgJson.devDependencies ?? {}) }

  if (packageManager) {
    targetPkgJson.packageManager = packageManager
  }

  for (const [depName, depVersion] of Object.entries(sourceDeps)) {
    if (typeof depVersion !== 'string') {
      continue
    }

    const targetVersion = targetDeps[depName]
    if (isWorkspace(targetVersion)) {
      continue
    }
    if (shouldAssignVersion(targetVersion, depVersion)) {
      targetDeps[depName] = depVersion
    }
  }
  if (Object.keys(targetDeps).length) {
    targetPkgJson.dependencies = targetDeps
  }

  for (const [depName, depVersion] of Object.entries(sourceDevDeps)) {
    if (typeof depVersion !== 'string') {
      continue
    }

    if (depName === pkgName) {
      const nextVersion = `^${pkgVersion}`
      const targetVersion = targetDevDeps[depName]
      if (!isWorkspace(targetVersion) && shouldAssignVersion(targetVersion, nextVersion)) {
        targetDevDeps[depName] = nextVersion
      }
    }
    else {
      const targetVersion = targetDevDeps[depName]
      if (isWorkspace(targetVersion)) {
        continue
      }
      if (shouldAssignVersion(targetVersion, depVersion)) {
        targetDevDeps[depName] = depVersion
      }
    }
  }
  if (Object.keys(targetDevDeps).length) {
    targetPkgJson.devDependencies = targetDevDeps
  }

  const scriptPairs = options?.scripts ? Object.entries(options.scripts) : scriptsEntries
  if (scriptPairs.length) {
    const scripts = { ...(targetPkgJson.scripts ?? {}) }
    for (const [scriptName, scriptCmd] of scriptPairs) {
      scripts[scriptName] = scriptCmd
    }
    targetPkgJson.scripts = scripts
  }
}
