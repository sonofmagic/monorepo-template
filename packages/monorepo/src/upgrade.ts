import type { Buffer } from 'node:buffer'
import type { CliOpts, PackageJson } from './types'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import checkbox from '@inquirer/checkbox'
import confirm from '@inquirer/confirm'
import defu from 'defu'
import fs from 'fs-extra'
import get from 'get-value'
import klaw from 'klaw'
import PQueue from 'p-queue'
import path from 'pathe'
import pc from 'picocolors'
import set from 'set-value'
import { name as pkgName, version as pkgVersion } from './constants'
import { logger } from './logger'
import { GitClient } from './monorepo/git'
import { scriptsEntries } from './scripts'
import { getAssetTargets } from './targets'
import { escapeStringRegexp, isFileChanged, isMatch } from './utils'

const queue = new PQueue({ concurrency: 1 })

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
// import.meta.dirname for Nodejs >= v20.11.0
// https://nodejs.org/api/esm.html#importmetadirname
const assetsDir = path.join(__dirname, '../assets')

function isWorkspace(version?: string) {
  if (typeof version === 'string') {
    return version.startsWith('workspace:')
  }
  return false
}

export function setPkgJson(sourcePkgJson: PackageJson, targetPkgJson: PackageJson) {
  const packageManager = get(sourcePkgJson, 'packageManager', { default: '' })
  const deps = get(sourcePkgJson, 'dependencies', { default: {} })
  const devDeps = get(sourcePkgJson, 'devDependencies', { default: {} })

  const targetDeps = get(targetPkgJson, 'dependencies', { default: {} })
  const targetDevDeps = get(targetPkgJson, 'devDependencies', { default: {} })

  set(targetPkgJson, 'packageManager', packageManager)
  Object.entries(deps).forEach((x) => {
    if (!isWorkspace(targetDeps[x[0]])) {
      set(targetPkgJson, `dependencies.${x[0].replaceAll('.', '\\.')}`, x[1], { preservePaths: false })
    }
  })
  Object.entries(devDeps).forEach((x) => {
    if (x[0] === pkgName) {
      set(targetPkgJson, `devDependencies.${x[0].replaceAll('.', '\\.')}`, `^${pkgVersion}`, { preservePaths: false })
    }
    else if (!isWorkspace(targetDevDeps[x[0]])) {
      set(targetPkgJson, `devDependencies.${x[0].replaceAll('.', '\\.')}`, x[1], { preservePaths: false })
    }
  })
  for (const [k, v] of scriptsEntries) {
    set(targetPkgJson, `scripts.${k}`, v)
  }
}

function confirmOverwrite(filename: string) {
  return confirm({ message: `${pc.greenBright(filename)} 文件内容发生改变,是否覆盖?`, default: false })
}

export async function upgradeMonorepo(opts: CliOpts) {
  const { outDir, raw, interactive, cwd } = defu<Required<CliOpts>, CliOpts[]>(opts, {
    cwd: process.cwd(),
    outDir: '',
  })
  const absOutDir = path.isAbsolute(outDir) ? outDir : path.join(cwd, outDir)
  const gitClient = new GitClient({
    baseDir: cwd,
  })
  const repoName = await gitClient.getRepoName()
  let targets = getAssetTargets(raw)

  if (interactive) {
    // https://github.com/pnpm/pnpm/blob/db420ab592666dbae77fdda3f5c04ed2c045846d/pkg-manager/plugin-commands-installation/src/update/index.ts
    targets = await checkbox({
      message: '选择你需要的文件',
      choices: targets.map((x) => {
        return {
          value: x,
          checked: true,
        }
      }),
    })
  }

  const regexpArr = targets.map((x) => {
    return new RegExp(`^${escapeStringRegexp(x)}`)
  })
  for await (const file of klaw(assetsDir, {
    filter(p) {
      const str = path.relative(assetsDir, p)
      return isMatch(str, regexpArr)
    },
  })) {
    await queue.add(async () => {
      if (file.stats.isFile()) {
        const relPath = path.relative(assetsDir, file.path)
        const targetPath = path.resolve(absOutDir, relPath)
        const targetIsExisted = await fs.exists(targetPath)

        async function overwriteOrCopy(target?: string | Buffer) {
          let isOverwrite = true
          if (targetIsExisted) {
            const src = await fs.readFile(file.path)
            const dest = target ?? await fs.readFile(targetPath)
            if (await isFileChanged(src, dest)) {
              isOverwrite = await confirmOverwrite(relPath)
            }
          }
          return isOverwrite
        }
        // const basename = path.basename(file.path)
        if (relPath === 'package.json') {
          const sourcePath = file.path
          if (targetIsExisted) {
            const sourcePkgJson = await fs.readJson(sourcePath) as PackageJson
            const targetPkgJson = await fs.readJson(targetPath) as PackageJson
            setPkgJson(sourcePkgJson, targetPkgJson)
            const data = JSON.stringify(targetPkgJson, undefined, 2)
            // packageJson
            // if (await overwriteOrCopy(data)) {
            await fs.writeFile(targetPath, `${data}\n`, 'utf8')
            logger.success(targetPath)
            // }
          }
        }
        else if (relPath === '.changeset/config.json' && repoName) {
          const changesetJson = await fs.readJson(file.path)
          set(changesetJson, 'changelog.1.repo', repoName)

          const data = JSON.stringify(changesetJson, undefined, 2)
          // changesetJson
          if (await overwriteOrCopy(data)) {
            await fs.ensureDir(path.dirname(targetPath))
            await fs.writeFile(targetPath, `${data}\n`, 'utf8')
            logger.success(targetPath)
          }
        }
        else if (await overwriteOrCopy()) {
          await fs.copy(
            file.path,
            targetPath,
          )
          logger.success(targetPath)
        }
      }
    })
  }
}
