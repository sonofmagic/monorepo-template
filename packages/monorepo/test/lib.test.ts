import type { PackageJson } from 'pkg-types'
import get from 'get-value'
import { version } from '@/constants'
import { setPkgJson } from '@/upgrade'

describe('lib', () => {
  it('setPkgJson casae 0', () => {
    const t: PackageJson = {
      packageManager: 'pnpm@9.12.1',
    }
    const o: PackageJson = {}
    setPkgJson(t, o)
    expect(o.packageManager).toBe('pnpm@9.12.1')
    expect(o).toMatchSnapshot()
  })

  it('setPkgJson casae 1', () => {
    const t = {

    }
    const o = {
      scripts: {
        'build': 'turbo run build',
        'dev': 'turbo run dev --parallel',
        'test': 'vitest run --coverage.enabled',
        'test:dev': 'vitest',
        'lint': 'turbo run lint',
        'release': 'changeset',
        'publish-packages': 'turbo run build lint test && changeset version && changeset publish',
        'preinstall': 'npx only-allow pnpm',
        'prepare': 'husky',
        'commit': 'commit',
      },
    }
    setPkgJson(t, o)
    expect(o).toMatchSnapshot()
  })

  it('setPkgJson casae 2', () => {
    const t: PackageJson = {
      dependencies: {
        xx: '0.0.0',
      },
      devDependencies: {
        x: '0.0.0',
      },
    }
    const o: PackageJson = {
      dependencies: {
        xx: 'workspace:*',
      },
      devDependencies: {
        x: 'workspace:*',
      },
    }
    setPkgJson(t, o)
    expect(o).toMatchSnapshot()
  })

  it('setPkgJson casae 3', () => {
    const t: PackageJson = {
      dependencies: {
        xx: '0.0.0',
      },
      devDependencies: {
        x: '0.0.0',
      },
    }
    const o: PackageJson = {
      dependencies: {
        xx: '2.2.2',
      },
      devDependencies: {
        x: '1.1.1',
      },
    }
    setPkgJson(t, o)
    expect(o).toMatchSnapshot()
  })

  it('setPkgJson casae 4', () => {
    const t: PackageJson = {
      devDependencies: {
        '@icebreakers/monorepo': '1',
      },
    }
    const o: PackageJson = {
      devDependencies: {
        '@icebreakers/monorepo': '2',
      },
    }
    setPkgJson(t, o)
    if (o.devDependencies) {
      expect(get(o.devDependencies, '@icebreakers/monorepo')).toBe(`^${version}`)
    }
  })
})
