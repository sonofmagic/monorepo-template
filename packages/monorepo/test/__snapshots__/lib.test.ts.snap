// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lib > setPkgJson casae 0 1`] = `
{
  "packageManager": "pnpm@9.12.1",
  "scripts": {
    "commitlint": "commitlint --edit",
    "script:clean": "monorepo clean",
    "script:init": "monorepo init",
    "script:mirror": "monorepo mirror",
    "script:sync": "monorepo sync",
  },
}
`;

exports[`lib > setPkgJson casae 1 1`] = `
{
  "packageManager": "",
  "scripts": {
    "build": "turbo run build",
    "commit": "commit",
    "commitlint": "commitlint --edit",
    "dev": "turbo run dev --parallel",
    "lint": "turbo run lint",
    "preinstall": "npx only-allow pnpm",
    "prepare": "husky",
    "publish-packages": "turbo run build lint test && changeset version && changeset publish",
    "release": "changeset",
    "script:clean": "monorepo clean",
    "script:init": "monorepo init",
    "script:mirror": "monorepo mirror",
    "script:sync": "monorepo sync",
    "test": "vitest run --coverage.enabled",
    "test:dev": "vitest",
  },
}
`;

exports[`lib > setPkgJson casae 2 1`] = `
{
  "dependencies": {
    "xx": "workspace:*",
  },
  "devDependencies": {
    "x": "workspace:*",
  },
  "packageManager": "",
  "scripts": {
    "commitlint": "commitlint --edit",
    "script:clean": "monorepo clean",
    "script:init": "monorepo init",
    "script:mirror": "monorepo mirror",
    "script:sync": "monorepo sync",
  },
}
`;

exports[`lib > setPkgJson casae 3 1`] = `
{
  "dependencies": {
    "xx": "0.0.0",
  },
  "devDependencies": {
    "x": "0.0.0",
  },
  "packageManager": "",
  "scripts": {
    "commitlint": "commitlint --edit",
    "script:clean": "monorepo clean",
    "script:init": "monorepo init",
    "script:mirror": "monorepo mirror",
    "script:sync": "monorepo sync",
  },
}
`;
